PR: Add Troubleshooting links and blurb to various appropriate places in Spyder and doc

CAM-Gerlach opened the pull request:
Adds a link and a short blurb regarding the new troubleshooting guide/FAQ in various appropriate places around the Spyder UI, including:

- About dialog
- Blurb/link as part of any crash/error dialog
- Blurb/link as part of the "crashed last time" startup message recommending spyder --reset
- Blurb/link in default Spyder issue template under .github/ISSUE_TEMPLATE.md
- And the auto-created/submitted one
- In the Readme, contributing.md, documentation index, and install doc
- Menu item in the help menu

Also some associated cleanup and polishing of related error and help guidance text, to help better guide users.

Finally, also includes a ``TROUBLESHOOTING.md`` document, along the lines of ``CONTRIBUTING.md`` which features a link and blurb to the full online guide/FAQ, as well as a few of the static sections that are likely to stay relatively unchanged, i.e, general troubleshooting tips and strategy, how to submit an issue, etc. Therefore, there is no need to hassle with maintaining it regularly, as it won't really become stale; and as its Markdown just like the wiki, it doesn't restrict compatibility whatsoever, and the relevant sections, if updated, can be easily copied and pasted in every once in a while if desired.

Thanks!

pep8speaks commented:
Hello @CAM-Gerlach! Thanks for submitting the PR.



 - In the file [`spyder/app/mainwindow.py`](https://github.com/spyder-ide/spyder/blob/621d274735bf5e8c559870006f03ca4cca090f11/spyder/app/mainwindow.py), following are the PEP8 issues :

> [Line 143:1](https://github.com/spyder-ide/spyder/blob/621d274735bf5e8c559870006f03ca4cca090f11/spyder/app/mainwindow.py#L143): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 941:41](https://github.com/spyder-ide/spyder/blob/621d274735bf5e8c559870006f03ca4cca090f11/spyder/app/mainwindow.py#L941): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 942:41](https://github.com/spyder-ide/spyder/blob/621d274735bf5e8c559870006f03ca4cca090f11/spyder/app/mainwindow.py#L942): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 2425:8](https://github.com/spyder-ide/spyder/blob/621d274735bf5e8c559870006f03ca4cca090f11/spyder/app/mainwindow.py#L2425): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented

ccordoba12 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -1,3 +1,9 @@
+**PLEASE READ:** Before posting this report, *please* carefully read our **Troubleshooting Guide** at https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ and search the issues page here for your error message and problem description, as the great majority of bugs are either duplicates, or can be fixed on the user side with a few easy steps.
```
Please break this at 80 characters or so, to not have it in a single line.

CAM-Gerlach replied:
I originally had it that way, but when it I tested how it would actually display on Github, it took up a a lot of excessive vertical space, since Github normally soft-wraps at 105 - 115 characters (at least in my browser) but respects the breaks in the file instead of rewrapping it properly and thus takes up ~50% more vertical room. 

I don't like leaving really long lines like this either, but pragmatically, as the intended use is for display on Github, and this file will normally never be seen/directly used (it only has one change in its entire history since the day it was first committed a few years ago, compared to several dozen issues created each day—around a ~5,000:1 difference), it seemed to make more sense to optimize for where it will actually matter 99.96% of the time. 

On the other hand, the narrower columns break nicely (except for the link, which I don't embed since its primarily for use in the raw/text view anyway) and help readability, It's not a big deal but just let me know which you prefer; you can compare them in the following screenshots illustrating the difference:

Editing View:

![image](https://user-images.githubusercontent.com/17051931/34631450-0a69189e-f236-11e7-854a-a0fe1c778587.png)
vs.
![image](https://user-images.githubusercontent.com/17051931/34631479-29100f6e-f236-11e7-8ca1-23df81f088c0.png)


Rendered View:

![image](https://user-images.githubusercontent.com/17051931/34632554-b567bf98-f23b-11e7-93dc-8facf2ba6c75.png)
vs.
![image](https://user-images.githubusercontent.com/17051931/34632588-eda3b51a-f23b-11e7-8a3f-b0ef4e6b273b.png)

ccordoba12 commented on .github/ISSUE_TEMPLATE.md:
```diff
@@ -1,3 +1,9 @@
+**PLEASE READ:** Before posting this report, *please* carefully read our **Troubleshooting Guide** at https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ and search the issues page here for your error message and problem description, as the great majority of bugs are either duplicates, or can be fixed on the user side with a few easy steps.
+
+Also, if you do post an issue here, please describe step by step what you were doing when the error occured, in as much detail as possible. Otherwise, we will have to close your issue after one week (7 days) if we can't reproduce it. Thanks.
```
Also break this single line in multiple ones.

CAM-Gerlach replied:
See above comment.

ccordoba12 commented on TROUBLESHOOTING.md:
```diff
@@ -0,0 +1,492 @@
+Spyder Troubleshooting Guide
```
Is this file referenced from somewhere else? If not, my vote is to remove it and only leave the one in the wiki.

The problem with having two places with the same content is that (inevitably) they will diverge at some point in time.

CAM-Gerlach replied:
> Is this file referenced from somewhere else? If not, my vote is to remove it and only leave the one in the wiki.

Fair enough; I figured this could be an objection, which is why I made it the last commit and not reference it elsewhere to make it easy to revert/remove if needed :) 

> The problem with having two places with the same content is that (inevitably) they will diverge at some point in time.

I wasn't going to include it for the same reason, but I ended up reconsidering it in the end after I specifically tailored it to only include the essentially static general troubleshooting content, which shouldn't change substantially at least over a reasonable timescale so it won't ever be that out of date, and it also includes a disclaimer and link at the top to the full/live version of the guide. Conveniently, the sections I included conveniently form the first and last chunk of the full version, and it uses the exact same Github markdown syntax, so all that needs to be done to update it every few months, if need be, is a simple copy and paste (which I am more than happy to do before each release, assuming there are actually any changes to the relevant sections). 

On the other hand, the benefits of having it there aren't enormous, but the main reason is that troubleshooting help is included with the program itself, same as the other documentation, readme, contributing, etc (which in some cases could also be argued to be duplicative, and much harder to maintain due to different syntax and structure), and so if the user doesn't have access to the Internet (or Github, in some countries), they are still able to get help. As a secondary benefit, it also makes it (and the linked guide) more visible/easily accessible to those browsing the source locally or on Github, and serves as a tracked, publicly accessible backup of the key sections in case Github wikis ever go down or go away.

However, I'm certainly far from wed to it, and already mentally prepared myself for the possible breakup :). With the above in mind, let me know what you'd like me to do with it. Thanks!

ccordoba12 commented:
@CAM-Gerlach, please post screenshots to see how this looks in all Spyder places you added this. Thanks!

CAM-Gerlach commented:
Sorry for the delay; just helped a friend get started using Spyder. I should have included them from the beginning, since already had to bring each one up anyway when checking that they displayed correctly. Here are the screenshots:

**About Spyder dialog:**
![image](https://user-images.githubusercontent.com/17051931/34635551-e69d8dfc-f256-11e7-947a-c5d93a8c6832.png)

**"Crashed last time" dialog:**
![image](https://user-images.githubusercontent.com/17051931/34635708-0adb4428-f259-11e7-9a06-a6d6426484d8.png)

**Error dialog:**
![image](https://user-images.githubusercontent.com/17051931/34635759-96b07afe-f259-11e7-8328-15e3a2f26460.png)

**Help Menu item:**
![image](https://user-images.githubusercontent.com/17051931/34635635-0496505e-f258-11e7-9924-7ed92d3338d7.png)

**Autogenerated error report**
![image](https://user-images.githubusercontent.com/17051931/34635775-dcfb9642-f259-11e7-8646-e98ab0295615.png)

Thanks!

ccordoba12 commented:
@CAM-Gerlach, this all looks very well! But since this involves adding new strings to the interface, I think I'll have to leave it for 3.2.7, sorry.

You could also help us to finish PR #5346, which would go well with your work here.

ccordoba12 commented:
I know you worked very hard to get this done in very short time, but the strings this PR changes are quite visible and so I'd prefer not to push them in a rush and leave them for 3.2.7 instead.

If you like, we could have that release by the end of the month with these changes, your PR about our README and the improvements to the report issues dialog I mentioned above. That would be a very good release too!

What do you say?

CAM-Gerlach commented:
> But since this PR involves adding new strings to the interface, I think I'll have to leave it for 3.2.7, sorry.

Okay...I assume for localization/internationalization? To be completely honest, I'm not quite sure how else one would expect to add a link and blurb to the UI...hieroglyphics? :P

CAM-Gerlach commented:
> You could also help us to finish PR #5346, which would go well with your work here.

Oh, nice! I was actually planning to add something like that anyway, but it didn't seem time-realistic nor really in scope for this PR.

ccordoba12 commented:
> Okay...I assume for localization/internationalization?

Yep, that's right.

> To be completely honest, I'm not quite sure how else one would expect to add a link and blurb to the UI...hieroglyphics? :P

If a string changes, then it falls back to English automatically. So our international users would see (for example) our Issues dialog in English. And that's not really useful :-)

CAM-Gerlach commented:
> If a string changes, then it falls back to English automatically. So our international users would see (for example) our Issues dialog in English. And that's not really useful :-)

Right, I understand that part, and we certainly don't want to leave Spyder's (many) international users out in the cold, especially on something important like this. Of course, the other issue is that the guide itself isn't translated...

What I was referring to was why it wasn't clear that this would be the case at some point before now. :-)

CAM-Gerlach commented:
> What do you say?

Sure, I guess. To be honest, while I was a bit miffed at first after pulling an all nighter (well, technically two) and calling in sick to work to make sure this got done, that was all my own (questionable) decisions, not anyone else's.  If there's one lesson I've learned in life as I've grown and matured, there is everything to be gained from accepting life as it comes and focusing on making the best of it. Though, in all fairness it would have been nice to know a *little* beforehand :-) 

One such possibility: Since the github issue template (both the manual and auto versions) is not localized, and that would by itself be able to (one would hope) be seen by at least all the users submitting issues, would you accept for 3.2.6 another PR with just that change?

Prikers commented:
I can do the French translation whenever you need me to, it should not take long. I mean for the new strings. For the troubleshooting guide itself it may take a bit more time, it is quite detailed :+1:

CAM-Gerlach commented:
> For the troubleshooting guide itself it may take a bit more time, it is quite detailed 

If you mean the FAQ version, then it will get much more detailed still...Soon™

So the readme, contributing.md, etc. are indeed translated too?

ccordoba12 commented:
> So the readme, contributing.md, etc. are indeed translated too?

Nop, they aren't. I don't know if that's a good idea though.

CAM-Gerlach commented:
Rebased and pared down to just the core UI commits, minus the report error dialog, which will be included in #6154 along with the github issue template and auto-report enhancements, and moved the doc additions and improvements to #6084 . Further, deconflicted with #5346 , implemented basic tests for the menu item and troubleshooting URL resolution, and did some additional minor cleanup. Should be ready to go on my end, unless you'd like very basic tests (essentially that they display and close properly, since they have no other specific functionality to test) for the "about" and "crashed last session" dialogs.
